src/utils/logger.ts

```diff
@@ -4,73 +4,90 @@ import { ISubmittableResult } from "@polkadot/types/types";
+            // signAndSend: Promise<() => void>
+            // signAndSend -> signAndSend resolves (we set unsubscribe) -> callback is called
+            transaction
+                .signAndSend(signer, { nonce: -1 }, (result: ISubmittableResult) => callback({ unsubscribe, result }))
+                .then((u: () => void) => (unsubscribe = u))
+                .catch((error) => reject(error));
+    
+            function callback(callbackObject: { unsubscribe: () => void; result: ISubmittableResult }): void {
+                // could log events here as they are being emitted
+                // using callbackObject.result.events
+                // noting that sometimes several events are
+                // emitted at once
+                const status = callbackObject.result.status;
+                if (status.isFinalized) {
+                    resolve(callbackObject);
                 }
             }
         });
-
-    if (!foundErrorEvent) {
-        events.forEach(({ phase, event: { data, method, section } }) => {
-            console.log(`\t' ${phase}: ${section}.${method}:: ${data}`);
-        });
-    } else if (errorMessage !== "issue.IssueCompleted") {
-        // IssueCompleted errors occur due to the vault having
-        // already auto-executed the issuance
-        throw new Error(errorMessage);
+    
+        console.log(`Transaction finalized at blockHash ${result.status.asFinalized}`);
+        unsubscribe(result);
+        this.printEvents(result.events);
+        return result;
     }
-}
-
-function isDispatchError(eventData: unknown): eventData is DispatchError {
-    return (eventData as DispatchError).isModule !== undefined;
+    
+    printEvents(events: EventRecord[]): void {
+        let foundErrorEvent = false;
+        let errorMessage = "";
+        events
+            .flatMap(({ event }) => event.data)
+            .forEach((eventData) => {
+                if (this.isDispatchError(eventData)) {
+                    try {
+                        const decoded = this.api.registry.findMetaError(eventData.asModule);
+                        const { documentation, name, section } = decoded;
+                        if (documentation && documentation.length > 0) {
+                            errorMessage = `${section}.${name}: ${documentation.join(" ")}`;
+                        } else {
+                            errorMessage = `${section}.${name}`;
+                        }
+                        foundErrorEvent = true;
+                    } catch (err) {
+                        errorMessage = "Error. Could not find transaction failure details.";
+                    }
+                }
+            });
+    
+        if (!foundErrorEvent) {
+            events.forEach(({ phase, event: { data, method, section } }) => {
+                console.log(`\t' ${phase}: ${section}.${method}:: ${data}`);
+            });
+        } else if (errorMessage !== "issue.IssueCompleted") {
```

Comment:
probably would be nice to have this as a constant in a file like events.ts. In case we would ever want to add more custom event error handling.